### PR TITLE
[not for land] CODA backward epilogue fusion prototype

### PR DESCRIPTION
**Pull request by soulitzer**
**Not for land - RFC / prototype.**

A UX prototype for CODA-style backward epilogue fusion across `autograd.Function`
nodes (see "CODA: Rewriting Transformer Blocks as GEMM-Epilogue Programs",
arXiv:2605.19269).

## Problem
The backward of an `autograd.Function` isn't self-contained: for a chain
`mm -> epilogue -> mm`, the epilogue's backward wants to fuse as the epilogue of
the *next* matmul's backward (`grad_a = (grad_c @ W^T) * f'(a)` is a matmul with a
pointwise epilogue), which a single Function can't express since its backward runs
inside one node. A main-backward GEMM followed by leaf gradient accumulation has
the same issue: eager materializes the GEMM result before `AccumulateGrad` adds it
to `.grad`. An implicitly multi-used activation has another boundary: autograd's
`InputBuffer` combines its branch gradients before the upstream node runs.

## Approach
The user writes a "1-to-many" op that decomposes into two autograd nodes (a matmul
node + an epilogue marker node), plus explicit `(pattern, replacement)` fusion
rules. Before backward, `apply_epilogue_fusion()` walks the graph and arms matching
main-backward outputs:

- For a main-backward -> epilogue-backward pair, the activation gradient is
  deferred into the previous epilogue's backward.
- For a main-backward -> `AccumulateGrad` pair, the output is deferred into an
  accumulator prehook, where the user's fused GEMM writes or adds directly into
  the leaf's `.grad`.
- For explicit multi-use, `split_multi_use(tensor, num_uses)` returns one alias
  per declared use. Distinct splitter output numbers give each branch its own
  `InputBuffer` slot. A pattern ending at `split_multi_use` replaces only the
  ordered branch-gradient accumulation and returns a real gradient to any upstream
  autograd node. A longer pattern can additionally replace an upstream CODA
  epilogue.

There is no graph rewrite: deferral rides a metadata-only placeholder along the
existing edge, carrying the producer's saved set while it is still alive. Leaf
accumulation fusion requires a unique producer so the engine never combines
placeholders. Leaf Tensor backward hooks bail out; post-accumulate hooks still run.
Multi-use fusion requires every declared splitter output to feed exactly one main
backward. A rule that also replaces an epilogue requires the splitter to be that
epilogue output's only consumer. Missing rules or invalid structures stay unfused,
and the splitter performs the ordinary sum; implicit multi-use without a splitter
continues to fail loudly.

## Files (top-level; the prototype lives out of tree)
- `coda_backward_epilogue_fusion.py` -- framework + example ops (MMRelu, MMTanh)
- `coda_backward_epilogue_fusion_test.py` -- device-generic numerics + planner tests
- `coda_backward_epilogue_fusion_lowlevel_demo.py` -- step-by-step demo of the plumbing

## Test plan
```
python coda_backward_epilogue_fusion.py
python coda_backward_epilogue_fusion_lowlevel_demo.py
python coda_backward_epilogue_fusion_test.py
```

Demos + 44 tests pass (CPU and CUDA). Coverage includes eager-equivalent numerics,
isolated activation-gradient accumulation at a split non-leaf native sigmoid,
end-to-end splitter accumulation composed with a CODA epilogue and leaf
accumulation, missing multi-use rules and invalid-use fallback, fresh and
pre-existing `.grad` (the latter retains its storage), activation and weight
outputs, shape-changing epilogues (GLU), shared-leaf / branching / missing-rule
bail-outs, post-accumulate hooks, a reference-cycle leak check, and composition
with `torch.utils.checkpoint(use_reentrant=False)` across two AC regions.

Authored with the assistance of AI coding assistants (Claude and OpenAI Codex).

### CODE REVIEW

File: coda_backward_epilogue_fusion.py [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+    captured at plan time because its real saved_tensors are freed before the
+    consumer's fused kernel runs. No pass-through wrapper is needed -- the fused
+    kernel reads only `saved_tensors` from the producer ctx."""
+
+    def __init__(self, saved):
+        self.saved_tensors = saved
+
+
+class _CtxWithNeedsInputGrad(_WrappedCtx):
+    """Override needs_input_grad -- to mark the deferred activation grad as not
```
Comment (albanD) [inline coda_backward_epilogue_fusion.py]: Not sure what this one means? :D

File: coda_backward_epilogue_fusion.py [hunk 2]
```diff
@@ -0,0 +1,654 @@
...
+            out = cls.forward(main_staging, epilogue_staging, *inputs)
+        # The intermediate (GEMM output) between the two nodes may differ in shape
+        # from `out` (e.g. SwiGLU is dim-reducing, norm involves a reduction), so the
+        # user must declare its metadata; we cannot infer it from `out`.
+        if main_staging.output_meta is None:
+            raise RuntimeError(
+                f"{cls.__name__}.forward must call main_ctx.set_output_meta(...) to "
+                f"declare the intermediate (main output) metadata"
+            )
+        # The per-call node classes form a Function<->Backward reference cycle, so
```
Comment (albanD) [inline coda_backward_epilogue_fusion.py]: Pretty insane trick haha

File: coda_backward_epilogue_fusion.py [hunk 3]
```diff
@@ -0,0 +1,654 @@
...
+def apply_epilogue_fusion(root, rules, expect_num_fusions=None):
+    r"""Plan backward epilogue fusion over the autograd graph reachable from ``root``.
+
+    Walks the backward graph and, for every adjacency where a producer op's
+    ``main_backward`` (a matmul backward) feeds a consumer op's
+    ``epilogue_backward`` (a pointwise backward), arms the producer to defer its
+    activation gradient so the consumer can run a single fused kernel instead --
+    the backward analogue of matmul epilogue fusion. The graph is **not** rewritten;
+    deferral is armed in place on the existing nodes, so this must be called after
+    forward and before :meth:`backward`, every iteration.
```
Comment (albanD) [inline coda_backward_epilogue_fusion.py]: Am I right that the tradeoff here is:
- The delayed Tensor is super simple as it is never worked on, as this pass ensures that it is only created when needed.
- There could be another design where we don't have this pass but the main would always create delayed Tensors that would get materialized whenever someone tries to use them.

We are leaning on the explicit rather an implicit here and simple vs easy. Is that right?

File: coda_backward_epilogue_fusion.py [hunk 4]
```diff
@@ -0,0 +1,654 @@
...
+
+    Example::
+
+        >>> import torch
+        >>> class MMRelu(FusibleFunction):
+        ...     @staticmethod
+        ...     def forward(main_ctx, epilogue_ctx, x, w):
+        ...         a = x @ w
+        ...         main_ctx.save_for_backward(x, w)
+        ...         main_ctx.set_output_meta(a)        # REQUIRED: boundary meta (else apply raises)
```
Comment (albanD) [inline coda_backward_epilogue_fusion.py:697]: nit: we could lift this and use the returned Tensor shape. With a nice error asking to use this if they don't match.
But explicit is ok for now

File: coda_backward_epilogue_fusion.py [hunk 5]
```diff
@@ -0,0 +1,654 @@
...
+        while q:
+            node = q.popleft()
+            for fn, _ in node.next_functions:
+                if fn is None or fn in seen:
+                    continue
+                seen.add(fn)
+                q.append(fn)
+            yield node
+
+    nodes = list(iter_graph([root]))
```
Comment (albanD) [inline coda_backward_epilogue_fusion.py]: Just build a list if you're not going to use the fact that it's a generator haha

File: coda_backward_epilogue_fusion.py [hunk 6]
```diff
@@ -0,0 +1,654 @@
...
+    in_degree = {}
+    for n in nodes:
+        for child, _ in n.next_functions:
+            if child is not None:
+                in_degree[child] = in_degree.get(child, 0) + 1
```
Comment (albanD) [inline coda_backward_epilogue_fusion.py]: This could be computed during the full traversal above?

File: coda_backward_epilogue_fusion.py [hunk 7]
```diff
@@ -0,0 +1,654 @@
...
+    (a_c,) = consumer_epilogue_ctx.saved_tensors
+    grad_main_input = grad_producer_out @ w_p.transpose(-1, -2)
+    return grad_main_input * (1 - torch.tanh(a_c) ** 2)
+
+
+# One rule per (producer.main_backward, consumer.epilogue_backward) pair, passed
+# explicitly to apply_epilogue_fusion. The fused impl depends only on the
+# consumer's epilogue, so both producers reuse the same impl.
+RULES = [
+    (MMRelu.main_backward, MMRelu.epilogue_backward, mm_relu_fused_backward),
```
Comment (albanD) [inline coda_backward_epilogue_fusion.py]: You either need:
- assert defer_input_idx == 0
- Have this rule take a grad_input_idx

File: coda_backward_epilogue_fusion.py (hunk 2 above)
Comment (soulitzer) [reply coda_backward_epilogue_fusion.py]: Yup, it might be no longer necessary though now that I'm passing in cls in directly

File: coda_backward_epilogue_fusion.py (hunk 1 above)
Comment (soulitzer) [reply coda_backward_epilogue_fusion.py]: Should've left more context in the comment here, but the trickiness here is that the user writing main_backward needs to to be very aware of needs_input_grad in order to make sure dW is run eagerly rather than deferred. 

Use a wrapped ctx in order to avoid mutating the existing ctx which can be annoying to deal with (which we ran into when trying to make needs_input_grad working with  inputs=)

File: coda_backward_epilogue_fusion.py (hunk 3 above)
Comment (soulitzer) [reply coda_backward_epilogue_fusion.py]: Yeah since the Epilogue node ends up seeing this tensor, might as well handle it explicitly there. It's also important in the case of hooks to error explicitly.

File: coda_backward_epilogue_fusion.py (hunk 4 above)
Comment (soulitzer) [reply coda_backward_epilogue_fusion.py:697]: I think the annoying thing here was that since the user has a fused_forward, although they might return this "a" intermediate for the purpose of saving for backward, its not obvious which output_idx it is, so it seems easiest to ask the user for the exact metadata so we know what placeholder tensor to thread from MainNode forward to EpilogueNode forward.

File: coda_backward_epilogue_fusion.py (hunk 5 above)
Comment (soulitzer) [reply coda_backward_epilogue_fusion.py]: Yeah this was dumb lol

File: coda_backward_epilogue_fusion.py (hunk 7 above)
Comment (soulitzer) [reply coda_backward_epilogue_fusion.py]: I think the current state is that we allow any input_idx and error if there's more than one possible consumer/Epilogue
